File: examples/leetcode-out/c/11/container-with-most-water.c
#include <stdio.h>
#include <stdlib.h>

typedef struct { int len; int *data; } list_int;

static list_int list_int_create(int len) {
	list_int l;
	l.len = len;
	l.data = (int*)malloc(sizeof(int)*len);
	return l;
}


int maxArea(list_int height){
	int left = 0;
	int right = (height.len - 1);
	int maxArea = 0;
	while ((left < right)) {
		int width = (right - left);
		int h = 0;
		if ((height.data[left] < height.data[right])) {
			h = height.data[left];
		} else {
			h = height.data[right];
		}
		int area = (h * width);
		if ((area > maxArea)) {
			maxArea = area;
		}
		if ((height.data[left] < height.data[right])) {
			left = (left + 1);
		} else {
			right = (right - 1);
		}
	}
	return maxArea;
}

int main() {
	return 0;
}

File: examples/leetcode-out/clj/11/container-with-most-water.clj
(ns main)

(defn _indexList [xs i]
  (let [idx (if (neg? i) (+ i (count xs)) i)]
    (if (or (< idx 0) (>= idx (count xs)))
      (throw (ex-info "index out of range" {}))
      (nth xs idx))))

(defn maxArea [height]
  (try
    (def left 0)
    (def right (- (count height) 1))
    (def maxArea 0)
    (loop []
      (when (< left right)
        (let [r (try
          (def width (- right left))
          (def h 0)
          (if (< (_indexList height left) (_indexList height right))
            (do
              (def h (_indexList height left))
            )
          
          (do
            (def h (_indexList height right))
          )
          )
          (def area (* h width))
          (when (> area maxArea)
            (def maxArea area)
          )
          (if (< (_indexList height left) (_indexList height right))
            (do
              (def left (+ left 1))
            )
          
          (do
            (def right (- right 1))
          )
          )
          :next
        (catch clojure.lang.ExceptionInfo e
          (cond
            (= (.getMessage e) "continue") :next
            (= (.getMessage e) "break") :break
            :else (throw e))
          )
        )]
      (cond
        (= r :break) nil
        (= r :next) (recur)
      )
    )
  )
)
(throw (ex-info "return" {:value maxArea}))
(catch clojure.lang.ExceptionInfo e
(if (= (.getMessage e) "return")
  (:value (ex-data e))
(throw e)))
)
)

(defn test_example_1 []
(assert (= (maxArea [1 8 6 2 5 4 8 3 7]) 49))
)

(defn test_example_2 []
(assert (= (maxArea [1 1]) 1))
)

(defn test_decreasing_heights []
(assert (= (maxArea [4 3 2 1 4]) 16))
)

(defn test_short_array []
(assert (= (maxArea [1 2 1]) 2))
)

(defn -main []
(test_example_1)
(test_example_2)
(test_decreasing_heights)
(test_short_array)
)

(-main)
